File: pconvert_rust/test/blending.js
const fs = require("fs");
const path = require("path");
const assert = require("assert");

const pconvert = require("../../pkg/pconvert_rust.js");

const TEST_ASSETS = path.resolve("assets/test");

describe("NodeJS WASM", async function() {
    this.timeout(30000);

    describe("#getModuleConstants", function() {
        it("should have known module constants", () => {
            const constants = pconvert.getModuleConstants();
            const mandatoryKeys = [
                "ALGORITHMS",
                "COMPILATION_DATE",
                "COMPILATION_TIME",
                "COMPILER",
                "COMPILER_VERSION",
                "COMPRESSION_TYPES",
                "FEATURES",
                "FILTER_TYPES",
                "LIBPNG_VERSION",
                "PLATFORM_CPU_BITS",
                "VERSION"
            ];
            assert.deepStrictEqual(mandatoryKeys, Object.keys(constants));
        });
    });

    describe("#blendMultipleFs", function() {
        it("should blend multiple files from local file system", () => {
            const paths = [
                path.resolve(`${TEST_ASSETS}/sole.png`),
                path.resolve(`${TEST_ASSETS}/back.png`),
                path.resolve(`${TEST_ASSETS}/front.png`),
                path.resolve(`${TEST_ASSETS}/shoelace.png`),
                path.resolve(`${TEST_ASSETS}/background_alpha.png`)
            ];
            const out = path.resolve(`${TEST_ASSETS}/result_alpha_alpha_Fast_NoFilter.png`);
            const algorithm = "alpha";
            const algorithms = ["alpha", "multiplicative", "destination_over", "source_over"];
            pconvert.blendMultipleFs(paths, out, algorithm, algorithms, true);
            assert(fs.existsSync(out));
        });

        it("should blend multiple files from local file system with no algorithm specified", () => {
            const paths = [
                path.resolve(`${TEST_ASSETS}/sole.png`),
                path.resolve(`${TEST_ASSETS}/back.png`),
                path.resolve(`${TEST_ASSETS}/front.png`),
                path.resolve(`${TEST_ASSETS}/shoelace.png`),
                path.resolve(`${TEST_ASSETS}/background_alpha.png`)
            ];
            const out = path.resolve(`${TEST_ASSETS}/result_alpha_alpha_Fast_NoFilter.png`);
            const algorithm = null;
            const algorithms = null;
            pconvert.blendMultipleFs(paths, out, algorithm, algorithms, true);
            assert(fs.existsSync(out));
        });
    });

    describe("#blendMultipleFsAsync", function() {
        it("should asynchronously blend multiple files from local file system", async () => {
            const paths = [
                path.resolve(`${TEST_ASSETS}/sole.png`),
                path.resolve(`${TEST_ASSETS}/back.png`),
                path.resolve(`${TEST_ASSETS}/front.png`),
                path.resolve(`${TEST_ASSETS}/shoelace.png`),
                path.resolve(`${TEST_ASSETS}/background_alpha.png`)
            ];
            const out = path.resolve(`${TEST_ASSETS}/result_alpha_alpha_Fast_NoFilter.png`);
            const algorithm = "alpha";
            const algorithms = ["alpha", "multiplicative", "destination_over", "source_over"];
            await pconvert.blendMultipleFsAsync(paths, out, algorithm, algorithms, true);
            assert(fs.existsSync(out), `Expected final composition to be at ${out}`);
        });

        it("should asynchronously blend multiple files from local file system with no algorithm specified", async () => {
            const paths = [
                path.resolve(`${TEST_ASSETS}/sole.png`),
                path.resolve(`${TEST_ASSETS}/back.png`),
                path.resolve(`${TEST_ASSETS}/front.png`),
                path.resolve(`${TEST_ASSETS}/shoelace.png`),
                path.resolve(`${TEST_ASSETS}/background_alpha.png`)
            ];
            const out = path.resolve(`${TEST_ASSETS}/result_alpha_alpha_Fast_NoFilter.png`);
            const algorithm = null;
            const algorithms = null;
            await pconvert.blendMultipleFsAsync(paths, out, algorithm, algorithms, true);
            assert(fs.existsSync(out), `Expected final composition to be at ${out}`);
        });
    });
});

File: pconvert_rust/test/blending.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import unittest
import pconvert_rust

TEST_ASSETS = os.path.join(os.path.dirname(__file__), "../../assets/test/")

class BlendingTest(unittest.TestCase):

    def test_module_constants(self):
        mandatory = [
            "ALGORITHMS",
            "COMPILATION_DATE",
            "COMPILATION_TIME",
            "COMPILER",
            "COMPILER_VERSION",
            "COMPRESSION_TYPES",
            "FEATURES",
            "FILTER_TYPES",
            "LIBPNG_VERSION",
            "PLATFORM_CPU_BITS",
            "VERSION"
        ];
        for key in mandatory: self.assertTrue(hasattr(pconvert_rust, key))

    def test_blend_images(self):
        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}sole.png"),
            os.path.abspath(f"{TEST_ASSETS}back.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png")
        )

        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            os.path.abspath(f"{TEST_ASSETS}front.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
        )

        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            os.path.abspath(f"{TEST_ASSETS}shoelace.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
        )

        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            os.path.abspath(f"{TEST_ASSETS}background_alpha.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
        )

    def test_blend_images_multithread(self):
        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}sole.png"),
            os.path.abspath(f"{TEST_ASSETS}back.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            options = {
                "num_threads": 5
            }
        )

        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            os.path.abspath(f"{TEST_ASSETS}front.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            options = {
                "num_threads": 5
            }
        )

        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            os.path.abspath(f"{TEST_ASSETS}shoelace.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            options = {
                "num_threads": 5
            }
        )

        pconvert_rust.blend_images(
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            os.path.abspath(f"{TEST_ASSETS}background_alpha.png"),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            options = {
                "num_threads": 5
            }
        )

    def test_blend_multiple(self):
        # default algorithms
        pconvert_rust.blend_multiple(
            (
                os.path.abspath(f"{TEST_ASSETS}sole.png"),
                os.path.abspath(f"{TEST_ASSETS}back.png"),
                os.path.abspath(f"{TEST_ASSETS}front.png"),
                os.path.abspath(f"{TEST_ASSETS}shoelace.png"),
                os.path.abspath(f"{TEST_ASSETS}background_alpha.png")
            ),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png")
        )

        # explicit algorithms
        pconvert_rust.blend_multiple(
            (
                os.path.abspath(f"{TEST_ASSETS}sole.png"),
                os.path.abspath(f"{TEST_ASSETS}back.png"),
                os.path.abspath(f"{TEST_ASSETS}front.png"),
                os.path.abspath(f"{TEST_ASSETS}shoelace.png"),
                os.path.abspath(f"{TEST_ASSETS}background_alpha.png")
            ),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            algorithm = "alpha",
            algorithms = ["alpha", "multiplicative", "destination_over", "source_over"]
        )

    def test_blend_multiple_multithread(self):
        # default algorithms
        pconvert_rust.blend_multiple(
            (
                os.path.abspath(f"{TEST_ASSETS}sole.png"),
                os.path.abspath(f"{TEST_ASSETS}back.png"),
                os.path.abspath(f"{TEST_ASSETS}front.png"),
                os.path.abspath(f"{TEST_ASSETS}shoelace.png"),
                os.path.abspath(f"{TEST_ASSETS}background_alpha.png")
            ),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            options = {
                "num_threads": 5
            }
        )

        # explicit algorithms
        pconvert_rust.blend_multiple(
            (
                os.path.abspath(f"{TEST_ASSETS}sole.png"),
                os.path.abspath(f"{TEST_ASSETS}back.png"),
                os.path.abspath(f"{TEST_ASSETS}front.png"),
                os.path.abspath(f"{TEST_ASSETS}shoelace.png"),
                os.path.abspath(f"{TEST_ASSETS}background_alpha.png")
            ),
            os.path.abspath(f"{TEST_ASSETS}result_alpha_alpha_Fast_NoFilter.png"),
            algorithm = "alpha",
            algorithms = ["alpha", "multiplicative", "destination_over", "source_over"],
            options = {
                "num_threads": 5
            }
        )
